site/content/blog/2020-12-04-common-i18n-mistakes.md

```diff
@@ -0,0 +1,315 @@
+
+### Solution 1: React elements as values
+
+Similar to the example on pluralization, we can actually pass translated text into another block of translated text as a value to construct the sentence.
+
+```jsx
+return (
+    <FormattedMessage
+        id='popup.moreInformation'
+        defaultMessage='{link} for more information.'
+        values={{
+            link: (
+                <a
+                    href='https://mattermost.com'
+                    rel='noreferrer'
+                    target='_blank'
+                >
+                    <FormattedMessage
+                        id='popup.moreInformation.link'
+                        defaultMessage='Click here'
+                    />
+                </a>
+            ),
+        }}
+    />
+);
+```
+
+You'll often see this used in older Mattermost code when we're trying to format part of a string. It allows us to add more complex formatting within translated text without making it impossible to translate.
+
+It does, however, make it harder for translators to understand the meaning of the text. They need to spend time reconstructing the entire sentence which may be difficult. It also makes the code more complicated to read for the same reason.
+
+### Solution 2: FormattedHTMLMessage
+
+`FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a compoent provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.
+
+```jsx
+// import {FormattedHTMLMessage} from 'react-intl';
+
+return (
+    <FormattedHTMLMessage
+        id='popup.moreInformation'
+        defaultMessage='<a href="https://mattermost.com" rel="noreferrer" target="_blank">Click here</a> for more information.'
+    />
+);
+```
+
+As with the previous solution, this allows translators to reconstruct the sentence however they need, but it comes with a few serious downsides such as:
+1. Translators needed to understand HTML, including more complicated things like the extra parameters on the `a` tag.
+2. It added the chance that an incorrect translation could break part of the application by accidentally including malformed HTML within a string.
+3. It also had potential performance impacts since we were going outside of React's DOM manipulation optimizations by using raw HTML.
+4. We lost the ability to wrap parts of the string in React elements which can have more functionality, such as encapsulating the additional parameters passed into `a` tag.
+5. Since it used HTML, `FormattedHTMLMessage` wasn't available in `react-native` for use in our mobile apps.
+
+But despite these issues, it helped point us in the right direction since it lead to a custom component that is much nicer to work with.
```

Comment:
```suggestion
But despite these issues, it helped point us in the right direction since it led to a custom component that's much nicer to work with.
```